scanner/trivy/scanner.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2021 The Rode Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trivy
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/google/uuid"
+	"github.com/rode/collector-image-scanner/scanner"
+	rode "github.com/rode/rode/proto/v1alpha1"
+	"github.com/rode/rode/protodeps/grafeas/proto/v1beta1/common_go_proto"
+	"github.com/rode/rode/protodeps/grafeas/proto/v1beta1/discovery_go_proto"
+	"github.com/rode/rode/protodeps/grafeas/proto/v1beta1/grafeas_go_proto"
+	"github.com/rode/rode/protodeps/grafeas/proto/v1beta1/package_go_proto"
+	"github.com/rode/rode/protodeps/grafeas/proto/v1beta1/vulnerability_go_proto"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type trivyImageScanner struct {
+	logger  *zap.Logger
+	rode    rode.RodeClient
+	trivy   CommandWrapper
+	version string
+}
+
+func NewImageScanner(logger *zap.Logger, client rode.RodeClient, trivyWrapper CommandWrapper) scanner.ImageScanner {
+	return &trivyImageScanner{
+		logger: logger,
+		rode:   client,
+		trivy:  trivyWrapper,
+	}
+}
+
+func (t *trivyImageScanner) Init() error {
+	t.logger.Info("Discovering Trivy version")
+	version, err := t.trivy.Version()
+	if err != nil {
+		return err
+	}
+	t.logger.Info("Found Trivy version", zap.String("version", version.Version))
+	t.version = version.Version
+	return nil
+}
+
+func (t *trivyImageScanner) ImageScan(imageUri string) {
+	log := t.logger.Named("ImageScan").With(zap.String("imageUri", imageUri))
+	log.Info("Starting scan")
+
+	results, err := t.trivy.Scan(imageUri)
+	if err != nil {
+		log.Error("Error scanning image", zap.Error(err))
+		return
+	}
+	log.Debug("Scan completed", zap.Duration("scan", results.ScanEnd.Sub(results.ScanStart)))
+
+	ctx := context.Background()
```

Comment:
we should probably take this context as an argument to the `ImageScan` method instead of creating it here.  When `ImageScan` is used by the `StartImageScan` method, we can take the context provided to that method and pass it on to `ImageScan`.  this will allow consumers of this method to cancel the request and propagate it down to `ImageScan`.

Reply:
I don't think we can take the context from the request. We spawn a goroutine in the RPC handler to kick the scan off, and that context is canceled at the end of the request:

```
2021-09-16T15:03:41.131-0400	ERROR	TrivyImageScanner.ImageScan	trivy/scanner.go:74	Error creating scan note	{"imageUri": "curlimages/curl@sha256:6e0a786e3e5181df00eaf3a0a1749c18a6bb20b01c9bd192ea72176ce8a1c94b", "error": "rpc error: code = Canceled desc = context canceled"}
github.com/rode/collector-image-scanner/scanner/trivy.(*trivyImageScanner).ImageScan
	/Users/alex/Developer/liatrio/rode/collector-image-scanner/scanner/trivy/scanner.go:74
```

Reply:
Ah, good point, I didn't realize the request finishing would cause the context to cancel.